src/entity/posts/hooks.ts

```diff
@@ -0,0 +1,183 @@
+  userId?: string;
+  req?: Pick<FastifyRequest, 'ip'>;
+}
+
+/**
+ * Apply Vordr filtering hook to check if post should be shadow banned
+ */
+export const applyVordrHook = async <T extends DeepPartial<Post>>(
+  post: T,
+  context: PreparePostContext,
+): Promise<T> => {
+  const postContent = (post as DeepPartial<FreeformPost>).content; // Some post types have content
+  const shouldVordr = await checkWithVordr(
+    {
+      id: post.id || 'new-post',
+      type: VordrFilterType.Post,
+      title: post.title || undefined,
+      content: postContent,
+    },
+    context,
+  );
+
+  if (shouldVordr) {
+    return {
+      ...post,
+      banned: true,
+      showOnFeed: false,
+      flags: {
+        ...post.flags,
+        vordr: true,
+        banned: true,
+        showOnFeed: false,
+      },
+    };
+  }
+
+  return post;
+};
+
+/**
+ * Apply Vordr filtering hook for post updates
+ */
+export const applyVordrHookForUpdate = async <T extends DeepPartial<Post>>(
+  updates: T,
+  existingPost: Pick<Post, 'id' | 'flags'>,
+  context: PreparePostContext,
+): Promise<T> => {
+  // If post is already vordr'd, don't check again
+  if (existingPost.flags?.vordr) {
+    return updates;
+  }
+
+  // Check content with existing checkWithVordr if we have userId
+  const updatesContent = (updates as DeepPartial<FreeformPost>).content; // Some post types have content
+  const shouldVordr = await checkWithVordr(
+    {
+      id: existingPost.id,
+      type: VordrFilterType.Post,
+      title: updates.title || undefined,
+      content: updatesContent,
+    },
+    context,
+  );
+
+  if (shouldVordr) {
+    logger.info({ postId: existingPost.id }, 'Post will be vordr on update');
+
+    return {
+      ...updates,
+      showOnFeed: false,
+      metadataChangedAt: new Date(),
+      flags: {
+        ...updates.flags,
+        vordr: true,
+        showOnFeed: false,
+      },
+    };
+  }
+
+  return updates;
+};
+
+/**
+ * Remove special characters for deduplication purposes while preserving
+ * letters from all languages.
+ * Removes punctuation, symbols, emojis, but keeps Unicode letters and digits.
+ * This is more aggressive than the general removeSpecialCharacters function
+ * and is specifically designed for content deduplication.
+ */
+export const removeAllSpecialCharactersForDedup = (text: string): string => {
+  // \p{L} matches any Unicode letter (including Chinese, Arabic, Cyrillic, etc.)
+  return text.replace(/[^\p{L}]/gu, '');
+};
+
+/**
+ * Normalize content for deduplication by making it lowercase,
+ * trimming whitespace, and removing ALL special characters including emojis
+ */
+export const normalizeContentForDeduplication = (content: string): string => {
+  return removeAllSpecialCharactersForDedup(content.toLowerCase().trim());
+};
+
+/**
+ * Generate SHA-256 hash from normalized content
+ */
+export const generateContentHash = (content: string): string => {
+  const hash = createHash('sha256');
```

Comment:
seen Viktor recommended sha1, probably is good as well, we used it for hashing code snippets

Reply:
Think sha256 or newer is better. Usually faster and less chance for collision